lib/ABI/FunctionType-revng-c.cpp

```diff
@@ -0,0 +1,424 @@
+  template<typename RegType>
+  std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingRegisterArguments(const SortedVector<RegType> &Used);
+
+  /// Helper used for converting stack argument struct into
+  /// the c-style representation
+  ///
+  /// \param StackArgumentTypes The qualified type of the relevant part of
+  ///        the stack.
+  /// \param IndexOffset The index of the first argument (should be set to
+  ///        the number of register arguments).
+  ///
+  /// \return An ordered list of arguments.
+  std::optional<llvm::SmallVector<model::Argument, 8>>
+  tryConvertingStackArguments(model::QualifiedType StackArgumentTypes,
+                              size_t IndexOffset) const;
+
+  /// Helper used for converting return values to the c-style representation
+  ///
+  /// \param Used a set of registers confirmed to be in use by
+  ///        the function in question.
+  /// \param Allowed an ordered list of registers allowed to be used
+  ///        for returning values by the ABI.
+  /// \param PtrToCopyLocation an optional register used for pointing to
+  ///        return values in memory on some architectures. It is set to
+  ///        `model::Register::Invalid` if it's not applicable.
+  ///
+  /// \tparam RegisterType type of a register, should be deduced from
+  ///         the \ref Used.
+  ///
+  /// \return a qualified type if conversion is possible, `std::nullopt`
+  ///         otherwise.
+  template<typename RegisterType>
+  std::optional<model::QualifiedType>
+  tryConvertingReturnValue(const SortedVector<RegisterType> &UsedRegisters);
+};
+
+class Converter : public ConverterBase {
+public:
+  using ConverterBase::ConverterBase;
+  TypeDependencyManager manageTypeDependencies(model::Binary &Binary) const {
+    return TypeDependencyManager(ConverterBase::manageTypeDependencies(Binary));
+  }
+
+public:
+  /// Entry point for the `toCABI` conversion.
+  std::optional<model::TypePath> toCABI(const model::RawFunctionType &Function,
+                                        TupleTree<model::Binary> &Binary) const;
+};
+
+std::optional<model::TypePath>
+Converter::toCABI(const model::RawFunctionType &Function,
+                  TupleTree<model::Binary> &Binary) const {
+  revng_log(Log, "Converting a `RawFunctionType` to `CABIFunctionType`.");
+  revng_log(Log, "Function:\n" << serializeToString(Function));
+
+  if (!ABI.checkCompatibility(Function.Arguments(), Function.ReturnValues())) {
+    revng_log(Log,
+              "FAIL: the function is not compatible with `"
+                << model::ABI::getName(ABI.ABI()) << "`.");
+    return std::nullopt;
+  }
+
+  // A single place to conveniently store all the types that need to be added
+  // to the model in order for the new function prototype to be valid.
+  auto M = manageTypeDependencies(*Binary);
+
+  // Convert register arguments first.
+  auto RegArguments = M.tryConvertingRegisterArguments(Function.Arguments());
```

Comment:
But 80 lines 😞.

I thought really hard about this one and couldn't come up with anything better. Apart from, I guess, doing
```cpp
const auto &InputArgs = Function.Arguments();
```
but then it's also an abbreviation - it doesn't fit otherwise.

I'm not going to compromise on function name, it needs to be descriptive. We already have `StackArguments` in the same scope, so going with just `Arguments` is even more confusing.

Reply:
```c++
  std::optional<llvm::SmallVector<model::Argument, 8>> RegisterArguments;
  RegisterArguments = M.tryConvertingRegisterArguments(Function.Arguments());
```
Consider a `using` for the `SmallVector`.

Reply:
I went with just `Arguments` after all: https://github.com/revng/revng/pull/232/commits/3acf37bf0f65237a14d09aebbcb3795c0b04d361.